### PR TITLE
Support looking up CloudFormation stack by tags

### PR DESCRIPTION
**Pull request by cb372**
Add a param to `UpdateAmiCloudFormationParameterTask` to specify tags for looking up the CF stack. `Stack` and `Stage` are automatically added to the list of tag filters if they are not already present, so usually only the `App` tag will need to be specified.

This more flexible way of looking up the CF stack is needed by CAPI because some of our stacks don't use the `<stack>-my-cf-stack-<stage>` naming scheme. In particular we have some nested CF stacks with auto-generated random IDs appended to their names.

@sihil @adamnfish @LATaylor-guardian

### CODE REVIEW

File: magenta-lib/src/main/scala/magenta/deployment_type/AmiCloudFormationParameter.scala [hunk 1]
```diff
@@ -31,21 +40,35 @@ object AmiCloudFormationParameter extends DeploymentType {
...
-      val stageName = Some(parameters.stage.name).filter(_ => appendStageToCloudFormationStackName(pkg))
-      val cloudFormationStackNameParts = Seq(stackName, Some(cloudFormationStackName(pkg)), stageName).flatten
-      val fullCloudFormationStackName = cloudFormationStackNameParts.mkString("-")
+      // Based on the supplied params, decide whether to look up the CF stack by name or by tags
+      val cloudFormationStackLookupStrategy = {
+        val tagsMap = cloudFormationStackTags(pkg)
+        if (tagsMap.nonEmpty) {
+          val stackAndStage = Map("Stage" -> parameters.stage.name) ++ stack.nameOption.map("Stack" -> _)
+          val withStackAndStageTags = stackAndStage.foldLeft[Map[String, String]](tagsMap){
+            // Don't overwrite any Stack or Stage tags that the user has specified
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/deployment_type/AmiCloudFormationParameter.scala]: Could you achieve the same thing by starting with `stackAndStage` and then just adding `tagsMap`?

**Comment by philwills** [comment]
👍 👍 

Using tags makes a lot more sense to me.

File: magenta-lib/src/main/scala/magenta/deployment_type/AmiCloudFormationParameter.scala (hunk 1 above)
Comment (cb372) [inline magenta-lib/src/main/scala/magenta/deployment_type/AmiCloudFormationParameter.scala]: You are a genius!

**Comment by sihil** [comment]
I'd prefer it if this was consistent with the way the tags are looked up for ASGs (i.e. infer stack, app and stage from the `deploy.json` and deploy parameters).

I appreciate that this isn't ideal - but it works and I'd rather keep it consistent and review all of the tag behaviour later than create an inconsistent behaviour. It was in fact a conversation around adding exactly this feature that was the impetus for creating the Riff-Raff future meeting next week :)

**Comment by sihil** [comment]
Superseded by #385

**Pull request closed without merging**